### PR TITLE
Candy Machine v0.3 - Candy Guards

### PR DESCRIPTION
**Pull request by Funkatronics**
## Description
- candy Machine v0.3 - Adding support for Candy Guards
- https://linear.app/metaplex/project/candy-machine-v3-v03-f7196c0c01d3/DX

## Work Completed
- added support for creating Candy Guards
- added support for wrapping a candy guard onto a candy machine

## API Changes
- Create a Candy Guard:
  ```
  val candyGuard = candyMachineClient.createCandyGuard(
    guards: List<Guard>, 
    groups: Map<String, List<Guard>>,
    authority: PublicKey
  )
  ```
- Add a Candy Guard to a Candy Machine (wrap Candy Guard):
  ```
  candyMachineClient.wrapCandyGuard(
    candyGuard: CandyGuard, 
    candyMachine: PublicKey,
    authority: Account?
) 
  ```

## Testing
- 105/105 tests passing

### CODE REVIEW

File: lib/src/main/java/com/metaplex/lib/experimental/jen/candyguard/Accounts.kt [hunk 1]
```diff
@@ -2,21 +2,30 @@
 // Accounts
 // Metaplex
 //
-// This code was generated locally by Funkatronics on 2022-09-28
+// This code was generated locally by Funkatronics on 2022-11-01
 //
 @file:UseSerializers(PublicKeyAs32ByteSerializer::class)
```
Comment (ajamaica) [inline lib/src/main/java/com/metaplex/lib/experimental/jen/candyguard/Accounts.kt:7]: I did not know this was posible

File: lib/src/main/java/com/metaplex/lib/experimental/jen/candyguard/idl.kt [hunk 2]
```diff
@@ -9,7 +9,7 @@ package com.metaplex.lib.experimental.jen.candyguard
 
 val candyGuardJson = """
```
Comment (ajamaica) [inline lib/src/main/java/com/metaplex/lib/experimental/jen/candyguard/idl.kt:10]: We need to start removing this IDLs form code an use a reliable source

File: lib/src/main/java/com/metaplex/lib/modules/candymachines/models/CandyGuard.kt [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+    data class NftBurn(val requiredCollection: PublicKey): Guard
+    data class NftGate(val requiredCollection: PublicKey): Guard
+    data class NftPayment(val requiredCollection: PublicKey, val destination: PublicKey): Guard
+    data class RedeemedAmount(val maximum: Long): Guard
+    data class SolPayment(val lamports: Long, val destination: PublicKey): Guard
+    data class StartDate(val date: Long): Guard
+    data class ThirdPartySigner(val signerKey: PublicKey): Guard
+    data class TokenBurn(val amount: Long, val mint: PublicKey): Guard
+    data class TokenGate(val amount: Long, val mint: PublicKey): Guard
+    data class TokenPayment(
```
Comment (ajamaica) [inline lib/src/main/java/com/metaplex/lib/modules/candymachines/models/CandyGuard.kt:55]: This is nice way to use it.

**Review by ajamaica — APPROVED** [review]
This is great. The testing is so clean and reliable. Great work Marco.

File: lib/src/main/java/com/metaplex/lib/experimental/jen/candyguard/idl.kt (hunk 2 above)
Comment (Funkatronics) [reply lib/src/main/java/com/metaplex/lib/experimental/jen/candyguard/idl.kt:10]: agreed

**Pull request merged**